### PR TITLE
Updated the Developer Guide with instructions for contributing code.

### LINKED ISSUE #692: Add auto backport functionality to backport PRs for releases
Coming from meta issue [opensearch-project/opensearch-plugins#108](https://github.com/opensearch-project/opensearch-plugins/issues/108):

- [x] Add auto-backport PR creation Github action to make sure all the PRs are backported for a particular release. 
- [x] Add auto-delete for merged auto-backport PRs to clean up the branches created during auto-backport.
- [ ] Add documentation for the same.

Related documentation: https://github.com/opensearch-project/opensearch-plugins/blob/main/BACKPORT.md

### PR DESCRIPTION
**Pull request by dlvenable**
### Description

This change documents our approach to branches and backports. I wrote it from the perspective of what a new contributor might want to know about the process.
 
### Issues Resolved

Resolves #692 by documenting backporting.
 
### Check List
- [ ] New functionality includes testing.
- [x] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: docs/developer_guide.md [hunk 1]
```diff
@@ -104,6 +104,36 @@ opensearch-data-prepper:1.2.0-SNAPSHOT
 ```
 
 
+## Contributing your Code to Data Prepper
+
+When you wish to contribute a change to Data Prepper, please create a GitHub Pull Request (PR). You should create your PR to target `main`.
```
Comment (cmanning09) [inline docs/developer_guide.md:109]: Take it or leave it: `When you wish` sounds passive and doesn't acknowledge our guidelines or standards. I would recommend using `ready`. I would also avoid the use of create in the second sentence as it is redundant from the first sentence and the requirement isn't on creating but rather the PR itself. 

```
When you are ready to contribute a change to Data Prepper, please create a GitHub Pull Request (PR). Your PR should target `main`.
```

File: docs/developer_guide.md [hunk 2]
```diff
@@ -104,6 +104,36 @@ opensearch-data-prepper:1.2.0-SNAPSHOT
...
+
+The Data Prepper maintainers will review your PR and merge it once it is approved.
+
+Some changes contain bug fixes or security fixes. If you believe your
```
Comment (cmanning09) [inline docs/developer_guide.md:113]: take it or leave it: I would recommend the call to action to be definite. "read on" is open ended.
```
Some changes containing bug fixes or security fixes may be eligible for a patch release. If you believe your change should be a patch release, please see [Backporting](#Backporting)
```

File: docs/developer_guide.md [hunk 3]
```diff
@@ -104,6 +104,36 @@ opensearch-data-prepper:1.2.0-SNAPSHOT
...
+When you create a PR which targets `main` and need this change as a patch to a previous version
+of Data Prepper, use the auto backport GitHub Action. All you need to do is add the label
+`backport <version>` to your PR which is targeting `main`. After the PR is merged, the GitHub
+Action will create a new PR to cherry-pick those changes into the `<version>` branch.
+A Data Prepper maintainer will need to approve and merge the backported code into the target branch.
+
+The auto-generated PR will be on a branch named `backport/backport-<original PR number>-to-<version>`.
+
+In general, only security fixes and bug fixes need to be backported. However, near the time of
+an upcoming release, we may backport new features.
```
Comment (cmanning09) [inline docs/developer_guide.md:135]: I am concerned we end up having to supporting multiple versions beyond bug and security fixes. I would rather encourage users to upgrade to the latest version.

@dlvenable, can you help clarify this. Is this limited to minor versions or will we support backporting across major versions as well?  Can you provide an example scenario as to why we should explicitly support backporting new features?

**Review by cmanning09 — COMMENTED** [review]
Thanks @dlvenable , I have a few minor comments and one about backporting. I am interested to hear your thoughts.

File: docs/developer_guide.md (hunk 3 above)
Comment (dlvenable) [reply docs/developer_guide.md:135]: Thanks for the feedback. I think that you and I are mostly in agreement on supporting multiple versions. I think I can improve the wording here.

I do believe we may at times wish to backport to the last version in a major release. Say for example, after Data Prepper 2.0.0 is out, we may still need to provide patches to Data Prepper 1.4. But, only security and important bug fixes.

**Comment by dlvenable** [comment]
@cmanning09 , I pushed a change based on your feedback. Please let me know what you think about the change to the Backporting section.

File: docs/developer_guide.md [hunk 4]
```diff
@@ -104,6 +104,39 @@ opensearch-data-prepper:1.2.0-SNAPSHOT
...
+When you create a PR which targets `main` and need this change as a patch to a previous version
+of Data Prepper, use the auto backport GitHub Action. All you need to do is add the label
+`backport <version>` to your PR which is targeting `main`. After the PR is merged, the GitHub
+Action will create a new PR to cherry-pick those changes into the `<version>` branch.
+A Data Prepper maintainer will need to approve and merge the backported code into the target branch.
+
+The auto-generated PR will be on a branch named `backport/backport-<original PR number>-to-<version>`.
+
+Data Prepper supports patch releases only on the latest version (e.g. 2.1) and on the last version
+for the previous major release (e.g. 1.4 after 2.0 has been released). These releases are
```
Comment (dinujoh) [inline docs/developer_guide.md:135]: >Data Prepper supports patch releases only on the latest version (e.g. 2.1) and on the last version
for the previous major release (e.g. 1.4 after 2.0 has been released).

Are we only going to support backporting changes to last two release Major releases ? How about other older Major release ?

Comment (dinujoh) [inline docs/developer_guide.md:135]: >Data Prepper supports patch releases only on the latest version (e.g. 2.1) and on the last version
for the previous major release (e.g. 1.4 after 2.0 has been released).

Are we only going to support backporting changes to last two release Major releases ? How about other older Major release ?

File: docs/developer_guide.md [hunk 5]
```diff
@@ -104,6 +104,39 @@ opensearch-data-prepper:1.2.0-SNAPSHOT
...
+only for bug fixes or security fixes. Please use backports only for bug and security fixes
+and only for candidate releases. You can ask the Data Prepper maintainers about backporting a
+change in your PR.
```
[the remark this answers is not in the document]
Comment (dlvenable) [reply docs/developer_guide.md:138]: I'll clarify that they can ask in their main PR.

**Review by cmanning09 — APPROVED** [review]

File: docs/developer_guide.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (dlvenable) [reply docs/developer_guide.md]: @dinujoh , I updated the wording here.

File: docs/developer_guide.md (hunk 4 above)
Comment (dlvenable) [reply docs/developer_guide.md:135]: We are only supporting the last release from a major release series.

Our current roadmap has us releasing 1.4 and then 2.0. After 1.4 comes out, we will only make patches to 1.4. The 1.3. branch will be done. After 2.0 comes out we will still support 1.4 as well with patches. After 3.0 comes out, we will no longer support 1.4 at all. Does this answer your question? Is there something I should change in the documentation to clarify this?

**Review by dinujoh — APPROVED** [review]

**Review by cmanning09 — APPROVED** [review]

**Pull request merged**